api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto

```diff
@@ -238,6 +238,75 @@ message HttpConnectionManager {
     google.protobuf.BoolValue enabled = 3;
   }
 
+  // [#not-implemented-hide] Transformations that apply to path
+  // headers. Transformations are applied before any processing of requests by
+  // HTTP filters, routing, and matching. A policy determines whether Envoy will
+  // forward the normalized *:path* header or forward the original header
+  // *:path* header (unless a filter modifies it).
+  message PathNormalizationOptions {
+    // Defines a policy to apply normalization. Any path rewrites that the
+    // router performs (e.g. :ref:`regex_rewrite
+    // <envoy_api_field_config.route.v3.RouteAction.regex_rewrite>` or
+    // :ref:`prefix_rewrite
+    // <envoy_api_field_config.route.v3.RouteAction.prefix_rewrite>`) will apply
+    // to the *:path* header destined for the upstream.
+    enum NormalizationPolicy {
+      // Normalization applies internally before any processing of requests by
+      // HTTP filters, routing, and matching. Envoy will forward the normalized
+      // *:path* header. Regex rewrites by the router apply to the normalized header.
+      FORWARD = 0;
+
+      // Normalization only applies internally. The original *:path* header will
+      // be sent upstream *unless* the path has been mutated by a filter.
+      INTERNAL = 1;
+    }
+
+    // Should paths be normalized according to RFC 3986 before any processing of
+    // requests by HTTP filters or routing? For paths that fail this check,
+    // Envoy will respond with 400 to paths that are malformed. See
+    // `Normalization and Comparison
+    // <https://tools.ietf.org/html/rfc3986#section-6>`_ for details of
+    // normalization. Note that Envoy does not perform `case normalization
+    // <https://tools.ietf.org/html/rfc3986#section-6.2.2.1>`_
+    message NormalizePath {
```

Comment:
@adisuissa Do you think I should combine these messages?
e.g.
```
message NormalizationControl {
    NormalizationPolicy policy = 1;
    google.protobuf.BoolValue value = 2;
}
// RFC 3986
NormalizationControl normalize_path = 3;
// Adjacent slashes
NormalizationControl merge_slashes = 3;
```

Reply:
I agree that both can be merged into `NormalizationControl` type, but IIUC the default value for the `BoolValue value` needs to be different for path-normalization and for slash-merging, no?
Also, if you think that in the future, path normalization will have sub-policy settings (e.g., disabling percent encoding), then it's better to keep them separate, and add a comment.

Reply:
Defaults could always be tuned with the param in `PROTOBUF_GET_WRAPPED_OR_DEFAULT`
Hmm after thinking about it, maybe some need individual controls, but a vast majority won't. I suppose `normalize_path` may need sub-options (for windows path?), but that's the only example I can think of.

Reply:
Another option is to add a new enum value: `DISABLE` to `NormalizationPolicy`. IIUC, this keeps the same semantics, removes the BoolValue flag, and potentially makes it easier to configure.
But if additional sub-options are needed then specific messages also make sense.

Reply:
That is a great idea! Let me try that out, I'll push it if it looks good. 
My feeling is that sub-options can always be made in to separate options (see waf transformation knobs https://github.com/SpiderLabs/ModSecurity/wiki/Reference-Manual-(v2.x)#Transformation_functions)

Reply:
I think I need to wrap enums with a WKT though.